### PR TITLE
Added ability to choose System.Text.Json serializer

### PR DESCRIPTION
**Pull request by anthony-mi**

### CODE REVIEW

File: src/Program.fs [hunk 1]
```diff
@@ -422,19 +446,18 @@ let generate (configFile: string) =
...
-
+                let files =
+                    generatedFiles
+                    |> Seq.map
+                        (fun file -> createCompileXElement config.createProjectFile outputDirectoryName file)
                 let packageReferences = [
                     yield! packageReferences
                     yield MSBuildXElement.PackageReferenceInclude("Fable.SimpleHttp", "3.0.0")
                     yield MSBuildXElement.PackageReferenceInclude("Fable.SimpleJson", "3.21.0")
+                    yield MSBuildXElement.PackageReferenceInclude("Newtonsoft.Json", "12.0.2")
```
Comment (Zaid-Ajaj) [inline src/Program.fs:460]: Fable projects do not work with Newtonsoft.Json, Fable won't be able to compile this project. Please remove `Newtonsoft.Json` from the package references

File: src/Program.fs [hunk 2]
```diff
@@ -139,18 +157,23 @@ let readConfig (file: string) =
...
-                        output = fullOutputPath
-                        errorType = errorType
-                        target = target
-                        createProjectFile = createProjectFile
-                        overrideClientName = overrideClientName
-                        copyLocalLockFileAssemblies = copyLocalLockFileAssemblies
-                        emitMetadata = emitMetadata
-                    }
+                    if serializer = SerializerType.System && target = OutputTarget.Fable ||
+                       serializer = SerializerType.System && target = OutputTarget.Shared then
```
Comment (Zaid-Ajaj) [inline src/Program.fs:161]: if `target = shared` && `serializer = SerializerType.System` it is still valid: only the generated dotnet client will have STJ, not the fable one

File: build/Program.fs [hunk 3]
```diff
@@ -54,103 +52,44 @@ let publish() =
         then failwith "Publish failed"
 
 let buildCraftSchema() =
-    if Shell.Exec(Tools.dotnet, "run -p Snowflaqe.fsproj -- --config ../samples/craft-cms/snowflaqe.json --generate", path [ solutionRoot; "src" ]) <> 0 then
+    if Shell.Exec(Tools.dotnet, "run -f netcoreapp3.1 -- --config ../samples/craft-cms/snowflaqe.json --generate", path [ solutionRoot; "src" ]) <> 0 then
```
Comment (Zaid-Ajaj) [inline build/Program.fs:55]: Please don't use a specific target framework here. Why do you need this?

File: src/Snowflaqe.fsproj [hunk 4]
```diff
@@ -7,13 +7,10 @@
...
-    <Version>1.24.0</Version>
-    <PackageReleaseNotes>
-      Fix copyLocalLockFileAssemblies option.
-      Update Fable.Remoting.Json and SimpleJson to latest.
-      Lower required FSharp.Core version to v4.7.2
-    </PackageReleaseNotes>
+    <TargetFrameworks>netcoreapp3.1;netstandard2.0</TargetFrameworks>
+    <Version>1.25.0</Version>
+    <PackageReleaseNotes>Added support for System.Text.Json</PackageReleaseNotes>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
```
Comment (Zaid-Ajaj) [inline src/Snowflaqe.fsproj:13]: Not a big fan of BindingRedirects, why is this needed here? Please try to remove it from here

File: src/Types.fs [hunk 5]
```diff
@@ -4,9 +4,15 @@ open GraphQLParser.AST
...
-    | Shared
+    | Fable = 0
+    | FSharp = 1
+    | Shared = 2
+
+[<RequireQualifiedAccess>]
+type SerializerType =
+    | System = 0
+    | Newtonsoft = 1
+    | Unknown = 2
```
Comment (Zaid-Ajaj) [inline src/Types.fs:15]: `Unknown` shouldn't be an option here. Default is `Newtonsoft`, the other option should be `SystemTextJson`

**Review by Zaid-Ajaj — CHANGES_REQUESTED** [review]
Looks good! I believe this supersedes #46? First I will get #47 merged in, then we can see how to work with this one

File: src/Program.fs [hunk 6]
```diff
@@ -583,9 +609,10 @@ let generate (configFile: string) =
                 colorprintfn "✏️  Generating Fable $green[%s]" sharedFableDocument
 
                 let packageReferences = [
-                    MSBuildXElement.PackageReferenceUpdate("FSharp.Core", "4.7.2")
+                    MSBuildXElement.PackageReferenceUpdate("FSharp.Core", "5.0.1")
                     MSBuildXElement.PackageReferenceInclude("Fable.SimpleHttp", "3.0.0")
                     MSBuildXElement.PackageReferenceInclude("Fable.SimpleJson", "3.21.0")
+                    MSBuildXElement.PackageReferenceInclude("Newtonsoft.Json", "12.0.2")
```
Comment (Zaid-Ajaj) [inline src/Program.fs:615]: Same as above, Newtonsoft.Json isn't needed in Fable projects, that what `Fable.SimpleJson` is for

File: src/Program.fs [hunk 7]
```diff
@@ -583,9 +609,10 @@ let generate (configFile: string) =
                 colorprintfn "✏️  Generating Fable $green[%s]" sharedFableDocument
 
                 let packageReferences = [
-                    MSBuildXElement.PackageReferenceUpdate("FSharp.Core", "4.7.2")
+                    MSBuildXElement.PackageReferenceUpdate("FSharp.Core", "5.0.1")
```
Comment (Zaid-Ajaj) [inline src/Program.fs:612]: I intentionally made this v4.7.2 for best compatibility, please revert

File: src/Types.fs [hunk 8]
```diff
@@ -4,9 +4,15 @@ open GraphQLParser.AST
 
 [<RequireQualifiedAccess>]
 type OutputTarget =
-    | Fable
-    | FSharp
-    | Shared
+    | Fable = 0
```
Comment (Zaid-Ajaj) [inline src/Types.fs:7]: Why the enum values here with 0, 1, 2?

File: src/Snowflaqe.fsproj [hunk 9]
```diff
@@ -7,13 +7,10 @@
...
     <IsPackable>true</IsPackable>
     <RollForward>Major</RollForward>
-    <TargetFramework>net5.0</TargetFramework>
-    <Version>1.24.0</Version>
-    <PackageReleaseNotes>
-      Fix copyLocalLockFileAssemblies option.
-      Update Fable.Remoting.Json and SimpleJson to latest.
-      Lower required FSharp.Core version to v4.7.2
-    </PackageReleaseNotes>
+    <TargetFrameworks>netcoreapp3.1;netstandard2.0</TargetFrameworks>
```
Comment (Zaid-Ajaj) [inline src/Snowflaqe.fsproj:10]: Using `<TargetFramework>net5.0</TargetFramework>` should suffice, please revert. Unless there is some issue with `net5.0`?

File: build/Program.fs (hunk 3 above)
Comment (xperiandri) [reply build/Program.fs:55]: Because we have both `netstandard2.0` and `netcoreapp3.1` targets. And `dotnet` requires to specify which one we want to execute

Comment (xperiandri) [reply build/Program.fs:55]: Maybe will we extract that to a constant?

File: src/Types.fs (hunk 5 above)
Comment (xperiandri) [reply src/Types.fs:15]: Why not move to System as default? Do you have any specific reason to stay with Newtonsoft?

File: src/Types.fs (hunk 8 above)
Comment (xperiandri) [reply src/Types.fs:7]: In order to parse from string

Comment (xperiandri) [reply src/Types.fs:7]: Using standard .NET functionality

File: src/Snowflaqe.fsproj (hunk 9 above)
Comment (xperiandri) [reply src/Snowflaqe.fsproj:10]: We need `netstandard2.0` for MSBuild running on .NET Framework (used in Visual Studio)
And `netcoreapp3.1` is used for MSBuild Core

**Comment by xperiandri** [comment]
> Looks good! I believe this supersedes #46? First I will get #47 merged in, then we can see how to work with this one

There are fixes for unit tests here that I've mentioned in Slack. Would you like us to apply them to #46 so that you merge it?

File: src/Snowflaqe.fsproj (hunk 9 above)
Comment (Zaid-Ajaj) [reply src/Snowflaqe.fsproj:10]: Alright, I see. That's fine then

File: src/Types.fs (hunk 5 above)
Comment (Zaid-Ajaj) [reply src/Types.fs:15]: The converter used from `Fable.Remoting.Json` has added features specifically to be able handle edge cases of GraphQL types such as unions and interfaces as I have mentioned before (see my comment [here](https://github.com/Zaid-Ajaj/Snowflaqe/pull/40#issuecomment-823187772))

The library has been battle-tested in production for years now which is enough reason for me to keep it as default.

File: build/Program.fs (hunk 3 above)
Comment (Zaid-Ajaj) [reply build/Program.fs:55]: OK, I see. A constant sounds good

File: src/Types.fs (hunk 5 above)
Comment (xperiandri) [reply src/Types.fs:15]: OK

File: build/Snowflaqe.Build.fsproj [hunk 10]
```diff
@@ -15,8 +15,11 @@
...
     <PackageReference Include="Fake.Core.Environment" Version="5.20.4" />
     <PackageReference Include="Fake.Core.Target" Version="5.20.4" />
-    <ProjectReference Include="..\src\Snowflaqe.fsproj" />
-    <PackageReference Update="FSharp.Core" Version="5.0.1" />
+    <PackageReference Update="FSharp.Core" Version="5.0.0" />
   </ItemGroup>
 
+	<ItemGroup>
+		<ProjectReference Include="..\src\Snowflaqe.fsproj" />
+	</ItemGroup>
```
Comment (xperiandri) [inline build/Snowflaqe.Build.fsproj:23]: Looks like tabs instead of spaces. Use **Fix mixed tabs** extension

File: src/CodeGen.fs [hunk 11]
```diff
@@ -846,8 +848,10 @@ let createFile fileName modules =
     let qualfiedNameOfFile = QualifiedNameOfFile.QualifiedNameOfFile(Ident.Create fileName)
     ParsedImplFileInput.ParsedImplFileInput(fileName, false, qualfiedNameOfFile, [], [], modules, (false, false))
 
-let formatAst file =
-    formatAst (ParsedInput.ImplFile file)
+let private formatConfig = { FormatConfig.FormatConfig.Default with EndOfLine = EndOfLineStyle.CRLF; NewlineBetweenTypeDefinitionAndMembers = true; StrictMode = true }
```
Comment (xperiandri) [inline src/CodeGen.fs:851]: Format multiline, i.e. move each property set to new line

**Comment by xperiandri** [comment]
@Zaid-Ajaj lets youse string interpolation in code generation. `sprintf` usage is not readable. OK?

File: src/CodeGen.fs [hunk 12]
```diff
@@ -1072,40 +1076,130 @@ let sampleFSharpClientMember query queryName hasVariables = stringBuffer {
...
+      (if hasVariables then "input: " + queryName + ".InputVariables" else "")
+      ("\"\"\"\n" + addLines query + "\n            \"\"\"")
+      (if hasVariables then "{ query = query; variables = Some input }" else "{ query = query; variables = None }")
+      queryName
+
+      queryName
+      (if hasVariables then "input: " + queryName + ".InputVariables" else "")
+      queryName
+      (if hasVariables then " input" else "()")
+}
```
Comment (xperiandri) [inline src/CodeGen.fs:1125]: Try using string interpolation. I suppose we don't need `stringBuffer` here.
It is better to cache values via `let` prior to string interpolation.

File: src/CodeGen.fs [hunk 13]
```diff
@@ -1072,40 +1076,130 @@ let sampleFSharpClientMember query queryName hasVariables = stringBuffer {
...
+type GraphqlInput<'T> = { query: string; variables: Option<'T> }
+type GraphqlSuccessResponse<'T> = { data: 'T }
+type GraphqlErrorResponse = { errors: %s list }
+
+type %s(url: string, httpClient: HttpClient, options: JsonSerializerOptions) =
+    new(url: string, options: JsonSerializerOptions) = %s(url, new HttpClient(), options)
+    new(url: string) = %s(url, new HttpClient(), new JsonSerializerOptions())
+
+    %s""" projectName errorType clientName clientName clientName members
+    }
```
Comment (xperiandri) [inline src/CodeGen.fs:1170]: Try using string interpolation. I suppose we don't need stringBuffer here.
It is better to cache values via let prior to string interpolation.

File: src/CodeGen.fs [hunk 14]
```diff
@@ -1072,40 +1076,130 @@ let sampleFSharpClientMember query queryName hasVariables = stringBuffer {
...
 
-    new(url: string) = %s(url, new HttpClient())
+    new(url: string, settings: JsonSerializerSettings) = %s(url, new HttpClient(), settings)
 
-%s""" projectName errorType clientName clientName members
-}
+    new(url: string) = %s(url, new HttpClient(), new JsonSerializerSettings())
+
+    %s""" projectName errorType clientName clientName clientName members
+    }
```
Comment (xperiandri) [inline src/CodeGen.fs:1198]: Try using string interpolation. I suppose we don't need stringBuffer here.
It is better to cache values via let prior to string interpolation.

File: src/Program.fs (hunk 2 above)
Comment (xperiandri) [reply src/Program.fs:161]: We need to implement a corresponding converter for Fable similar to your converter to allow this scenario. That is why it is not allowed yet. @Zaid-Ajaj, am I wrong?

**Comment by Zaid-Ajaj** [comment]
> @Zaid-Ajaj lets youse string interpolation in code generation. sprintf usage is not readable. OK?

Sounds good to me!

File: src/Program.fs [hunk 15]
```diff
@@ -422,19 +446,18 @@ let generate (configFile: string) =
...
-
-                let files = [
-                    for file in generatedFiles do
-                        createCompileXElement config.createProjectFile outputDirectoryName file
-                ]
-
+                let files =
+                    generatedFiles
+                    |> Seq.map
+                        (fun file -> createCompileXElement config.createProjectFile outputDirectoryName file)
```
Comment (xperiandri) [inline src/Program.fs:455]:
```suggestion
                let files = seq {
                    for file in generatedFiles do
                        yield createCompileXElement config.createProjectFile outputDirectoryName file
                }
```

Let's go this way everywhere similar code exist

File: src/Program.fs [hunk 16]
```diff
@@ -527,21 +552,22 @@ let generate (configFile: string) =
...
                     then Path.GetFullPath(Path.Combine(config.output, "dotnet", config.project + ".Dotnet.fsproj"))
                     else Path.GetFullPath(Path.Combine(config.output, "dotnet", config.project + ".props"))
                 colorprintfn "✏️  Generating F# dotnet $green[%s]" sharedFSharpDocument
-
                 let packageReferences = [
                     yield! packageReferences
-                    yield MSBuildXElement.PackageReferenceInclude("Fable.Remoting.Json", "2.17.0")
+                    yield match config.serializer with
+                          | SerializerType.System -> MSBuildXElement.PackageReferenceInclude("System.text.Json", "4.6.0")
+                          | SerializerType.Newtonsoft -> MSBuildXElement.PackageReferenceInclude("Fable.Remoting.Json", "2.14.0")
```
Comment (xperiandri) [inline src/Program.fs:568]: Use the same version as below

**Review by xperiandri — CHANGES_REQUESTED** [review]

**Comment by Zaid-Ajaj** [comment]
> We need to implement a corresponding converter for Fable similar to your converter to allow this scenario. That is why it is not allowed yet. @Zaid-Ajaj, am I wrong?

for `serializer = system && target = fable` it is not allowed because Fable cannot use System.Text.Json and instead has `Fable.SimpleJson` 

for `serializer = system && target = shared` it _is_ allowed: shared project stays as is, generated fable project stays as is, only the generated dotnet project will then use System.Text.Json

File: src/Program.fs (hunk 1 above)
Comment (anthony-mi) [reply src/Program.fs:460]: @Zaid-Ajaj there are references to `Newtonsoft.Json` and `Newtonsoft.Json.Linq` in `GraphqlClient` which opened access to `JsonSerializerSettings`, `JsonConverter` etc.

Comment (Zaid-Ajaj) [reply src/Program.fs:460]: No, not for the Fable targets. These don't rely on Newtonsoft.Json

Comment (xperiandri) [reply src/Program.fs:460]: Why integration tests fail then?

Comment (Zaid-Ajaj) [reply src/Program.fs:460]: Which integration tests you mean?

**Comment by xperiandri** [comment]
@Zaid-Ajaj, @anthony-mi fixed all comments except the question with Newtonsoft.Json references

**Comment by Zaid-Ajaj** [comment]
I can take try fix the newtonsoft references issue myself but first, we have to resolve the merge conflict. Can you please rebase on master?

**Comment by Zaid-Ajaj** [comment]
This look good! thanks @anthony-mi and @xperiandri for the implementation. There is indeed some weird stuff with the newtonsoft references. I will merge this now and try to fix it

**Comment by Zaid-Ajaj** [comment]
I have just fixed all issues with package references and other small stuff, then published Snowflaqe v1.26 🚀 this version includes the `serializer` option. Can you please give it a try? 

I do have one issue though when `serializer = system`, the generated client type uses the _barebones_ System.Text.Json without any knowledge about how to deserialize F# types:
```fs
type SpotifyGraphqlClient(url: string, httpClient: HttpClient, options: JsonSerializerOptions) =
    new(url: string, options: JsonSerializerOptions) = SpotifyGraphqlClient(url, new HttpClient(), options)
    new(url: string) = SpotifyGraphqlClient(url, new HttpClient(), new JsonSerializerOptions())
```
Notice here, that by default we use `new JsonSerializerOptions()` when no options are provided. Shouldn't we include a specialized F# serializer from [FSharp.SystemTextJson](https://github.com/Tarmil/FSharp.SystemTextJson) by default as follows (from [docs](https://github.com/Tarmil/FSharp.SystemTextJson/blob/master/docs/Using.md#using-options)):
```fs
open System.Text.Json
open System.Text.Json.Serialization

let options = JsonSerializerOptions()
options.Converters.Add(JsonFSharpConverter())
```
I know that consumers of the library could do this themselves but I can imagine it would be better to do it from here so that the deserialization works out of the box without extra configuration. What do you think?

**Comment by xperiandri** [comment]
> Shouldn't we include a specialized F# serializer from [FSharp.SystemTextJson](https://github.com/Tarmil/FSharp.SystemTextJson) by default

Yes, we should. Currently, for Newtonsoft we always insert it to index 0 by default.
Maybe we need to align this implementation for both in a fashion where we insert it to index 0 only when it is not present? Or only in the case of default options?

**Comment by Zaid-Ajaj** [comment]
Only for default options.

To be honest, if we make sure the defaults _Just Work_ out of the box, I believe we don't even need to expose the JsonSerializerOptions to the consumers since they can just start using the client and everything will work like it should 🤞 

This is what I did for the default newtonsoft serializer: I removed the options from the constructor of the client because the defaults will work as is.

**Comment by Zaid-Ajaj** [comment]
Ah one more thing. Using target framework `netstandard2.0` was not allowed when packing a dotnet CLI, so I published latest for `netcoreapp3.1` and `net5.0`

**Comment by xperiandri** [comment]
Not good. You need to publish `netcoreapp3.1` only keeping `netstandard2.0` present in the library.
Because we need `netstandard2.0` for MSBuild

**Comment by xperiandri** [comment]
> Only for default options.

OK, we will change that

**Comment by Zaid-Ajaj** [comment]
To get the netstandard2.0 to work, I think I need to split the tool into two projects: a reusable library targeting netstandard2.0 and a CLI tool targeting netcoreapp3.1/net5.0

**Comment by xperiandri** [comment]
I suppose you can pass a switch to pack command

**Comment by xperiandri** [comment]
@Zaid-Ajaj so have you removed the ability to set `JsonSerializerSetting` for Newtosoft version of a client?

**Comment by xperiandri** [comment]
My consideration about it is that if you generate your code from the command line you can do whatever you want with sources then.
But if you add Snowflaqe from NuGet as MSBuild the code will be built into `obj` folder and regenerated in project rebuild. So that you will not be able to change anything.

**Comment by Zaid-Ajaj** [comment]
The thing is, I don't want people to change serialisation settings because what's included is guaranteed to work with these settings. If this is limiting and there is indeed a good reason for changing the settings, I am all ears for it

**Comment by xperiandri** [comment]
What about custom scalars?

**Comment by Zaid-Ajaj** [comment]
Custom scalars are converted as `string` except when they are _known_ custom scalars in which they are converted to specialized types

**Comment by xperiandri** [comment]
Ah, there is no mechanism to override that yet, right?

**Comment by Zaid-Ajaj** [comment]
That's correct

**Comment by xperiandri** [comment]
We need to add that 🙂
